File: src/easy/020-valid-parentheses/validParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  if (s.length <= 0 || s.length % 2 !== 0) {
    return false;
  };

  const map = {
    "(" : ")",
    "[" : "]",
    "{" : "}",
  };

  const stack = [];

  for (let i = 0; i < s.length; i++) {
    if (map[s[i]]) {
      stack.push(s[i]);
    } else {
      let brace = stack.pop();
      if (map[brace] !== s[i]) {
        return false;
      }
    }
  }

  return !stack.length;
};

console.log(isValid("()"));       // output = true
console.log(isValid("()[]{}"));   // output = true
console.log(isValid("(]"));       // output = false
console.log(isValid("(("));       // output = false

File: src/easy/020-valid-parentheses/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        size = len(s)

        if (size <= 0 or size % 2 != 0):
            return False

        hashmap = {
            "(": ")",
            "[": "]",
            "{": "}",
        }

        stack = []

        for _x in s:
            if _x in hashmap:
                stack.append(_x)
            else:
                if len(stack) == 0:
                    return False

                brace = stack.pop()
                if hashmap[brace] != _x:
                    return False

        return len(stack) == 0


testSolution = Solution()

print(testSolution.isValid("()"))       # output = True
print(testSolution.isValid("()[]{}"))   # output = True
print(testSolution.isValid("(]"))       # output = False
print(testSolution.isValid("(("))       # output = False
print(testSolution.isValid("){"))       # output == False
